[Png Yi Wei Jonas] ip

Jonaspng opened the pull request:
# Duke

>Give me a lever long enough and a fulcrum on which to place it, and I shall move the world. -Archimedes ([Source](https://www.brainyquote.com/quotes/archimedes_101761#:~:text=Archimedes%20Quotes&text=Please%20enable%20Javascript-,Give%20me%20a%20lever%20long%20enough%20and%20a%20fulcrum%20on,I%20shall%20move%20the%20world.))

   ```
   Hello from
    ____        _        
   |  _ \ _   _| | _____ 
   | | | | | | | |/ / _ \
   | |_| | |_| |   <  __/
   |____/ \__,_|_|\_\___|
   ```


Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

And it is **FREE**!

## How do I get it?

Prerequisites: Has Java downloaded on computer

All you need to do is,
1. Download from the Duke.jar file [here](https://github.com/Jonaspng/ip/releases/tag/A-Jar)
2. Go to command prompt and run Java -jar Duke.jar in the same directory as where Duke.jar file is located
3. Add your tasks.
4. Let it manage your tasks for you 😉

## Features

- [x] Managing Task
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

## For Java Programmers

If you Java programmer, you can use it to practice Java too. Here's the main method in Duke.java:

```
    public static void main(String[] args) {
        new Duke("./data").run();
    }
```

Commit:
Add Gradle support

Jonaspng committed:
Implemented Level 1

Jonaspng committed:
Added java docs

Jonaspng committed:
Implemented Level 2

Jonaspng committed:
changed : to . instead for item list

Jonaspng committed:
Implemented level 3

Jonaspng committed:
implemented level 4

Jonaspng committed:
implemented automated text ui testing

Jonaspng committed:
implemented exception handling

Jonaspng committed:
Implemented level 6

Jonaspng committed:
swtiched to using enum

Jonaspng committed:
Implemented level 7

Jonaspng committed:
Implement Date for deadline and events

Jonaspng committed:
Merge branch 'branch-Level-7'

Jonaspng committed:
Merge branch 'branch-Level-8'

Jonaspng committed:
Implement more classes for OOP

Jonaspng committed:
Implement duke package

Jonaspng committed:
Merge branch 'add-gradle-support' of https://github.com/Jonaspng/ip

Jonaspng committed:
Add gradle

Jonaspng committed:
Add junit tests

Jonaspng committed:
Change gradle settings for .jar file

Jonaspng committed:
Add java docs

Jonaspng committed:
Follow java coding standard

Jonaspng committed:
Implement java coding standard to java doc

Jonaspng committed:
Implement find task functionality

Jonaspng committed:
Merge branch 'branch-A-JavaDoc'

Jonaspng committed:
Merge branch 'branch-A-CodingStandard'

Jonaspng committed:
Merge branch 'branch-Level-9'

Jonaspng committed:
Add more java docs

Jonaspng committed:
Change TaskList to fit coding standard

Jonaspng committed:
Fix remove invalid task bug

Jonaspng committed:
Update Junit test

Jonaspng committed:
Implement CheckStyle

Jonaspng committed:
Revert "Implement CheckStyle"

This reverts commit ad91f7b30e4d61d1428eb128171852247a9095c0.

Jonaspng committed:
Implement CheckStyle

Jonaspng committed:
Implement Duke GUI

Jonaspng committed:
Merge branch 'branch-A-CheckStyle'

Jonaspng committed:
Merge branch 'branch-Level-10'

Jonaspng committed:
Change to fit coding standard

Jonaspng committed:
Update GUI

Jonaspng committed:
Resolve javafx setstyle parse error

Bacon-Strips commented on data/duke.txt:
```diff
@@ -0,0 +1,4 @@
+T | 1 | test
+E | 1 | testing | 2022-08-31
+D | 1 | assignment | 2022-08-31
+T | 0 | testing
```
Might want to edit the `.gitignore` file to ignore the save data files when pushing commits.

Jonaspng replied:
I will implement it soon. Thanks

Bacon-Strips commented on src/main/java/duke/AddCommand.java:
```diff
@@ -0,0 +1,34 @@
+ */
+public class AddCommand extends Command {
+
+    private Task newTask;
```
```suggestion
    private final Task newTask;
```
Suggestion for the addition of `final` modifier.

Jonaspng replied:
I did not implement it as final as I may need to modify the said variable later on.

Bacon-Strips commented on src/main/java/duke/DeleteCommand.java:
```diff
@@ -0,0 +1,35 @@
+ */
+public class DeleteCommand extends Command {
+
+    private int taskNumber;
```
```suggestion
    private final int taskNumber;
```
Suggesting the addition of `final` modifier.

Jonaspng replied:
I did not implement it as final as I may need to modify the said variable later on.

Bacon-Strips commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public class Duke {
+
+    private Storage storage;
+    private TaskList tasks;
+    private Ui ui;
```
```suggestion
    private final Storage storage;
    private final TaskList tasks;
    private final Ui ui;
```
Suggesting the addition of `final` modifier.

Jonaspng replied:
I did not implement it as final as I may need to modify the said variable later on.

Bacon-Strips commented on src/main/java/duke/FindCommand.java:
```diff
@@ -0,0 +1,32 @@
+ */
+public class FindCommand extends Command {
+
+    private String taskDescription;
```
```suggestion
    private final String taskDescription;
```
Suggesting the addition of `final` modifier.

Jonaspng replied:
I did not implement it as final as I may need to modify the said variable later on.

Bacon-Strips commented on src/main/java/duke/MarkCommand.java:
```diff
@@ -0,0 +1,30 @@
+ */
+public class MarkCommand extends Command {
+
+    private int taskNumber;
```
```suggestion
    private final int taskNumber;
```
Suggesting the addition of `final` modifier.

Jonaspng replied:
I did not implement it as final as I may need to modify the said variable later on.

Bacon-Strips reviewed (commented):
Your code seems well done. Looks like you implemented a decent amount of tests too.

Jonaspng committed:
Add assert statements

weesiau commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+ */
+public class Deadline extends Task {
+
+    protected LocalDate by;
```
```suggestion
    protected final LocalDate by;
```
Should you consider setting this attribute as final?

Jonaspng replied:
I did not implement it as final as I may need to modify the date later on. For example, if a user wants to change the by date in deadline.

weesiau commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,29 @@
+ */
+public abstract class Command {
+
+    private boolean isExit = false;
```
Can this name be more descriptive?

Jonaspng replied:
Hmm. I took isExit from the CS2103T website though?

weesiau commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,45 @@
+ */
+public class Event extends Task {
+
+    protected LocalDate at;
```
```suggestion
    protected final LocalDate at;
```
Should this attribute be set as final?

Jonaspng replied:
I did not implement it as final as I may need to modify the date later on. For example, if a user wants to change the at date in event.

weesiau commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,65 @@
+    private void handleUserInput() {
+        String response = "";
+        String input = "";
+        String color = "#6adcd3";
+        try {
+            input = userInput.getText();
+            response = duke.getResponse(input);
+        } catch (DukeException e) {
+            color = "#ffb6c1";
```
```suggestion
        String final COLOR_BLUE = "#6adcd3";
        String final COLOR_RED = "#ffb6c1"
        String color = COLOR_BLUE;
        try {
            input = userInput.getText();
            response = duke.getResponse(input);
        } catch (DukeException e) {
            color = COLOR_RED;
```
Should the colour codes be stored as descriptively named final attributes for better understanding?

Jonaspng replied:
Thats a good idea! Thanks 😄

weesiau commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,129 @@
+     */
+    public TaskList find(String userInput) {
+        ArrayList<Task> currTasks = new ArrayList<>(this.tasks);
+        currTasks.removeIf(x -> !x.getDescription().contains(userInput));
```
I like the usage of removeIf here

Jonaspng replied:
You should try using it too. It cuts down the amount of code you need to implement if you use filter instead.

weesiau reviewed (commented):
Overall, I found the code very compliant with the coding standard and easy to read, with clever tricks used throughout. However some parts can be better described, be it by better variable naming or different handling of certain strings like colour codes.

Jonaspng committed:
Update build.gradle to enable assertions

Vshnv2001 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,43 @@
+package duke;
```
Suggesting the separation of your classes into multiple packages instead of a sole Duke package that has multiple classes- allows for better organization of code

Jonaspng replied:
I will take your suggestion into consideration

Vshnv2001 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+     * @return string that represent this deadline object, to be stored in data file.
+     */
+    @Override
+    public String toStorageString() {
```
I like your use of inheritance in the toStorageString method

Jonaspng replied:
Thanks.

Vshnv2001 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,43 @@
+        boolean isExit = false;
+        String dukeResponse = "";
+        Command c = Parser.parse(userInput);
+        dukeResponse = c.execute(tasks, ui, storage);
```
Suggesting the addition of the 'this' keyword to identify variables native to an object

Jonaspng replied:
I will take your comment into consideration.

Vshnv2001 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,155 @@
+     */
+    public static Command parse(String input) throws DukeException {
+        String[] inputList = input.split(" ");
+        Task newListItem;
```
Your use of switch case is good! Makes your code more readable and neater

Jonaspng replied:
Thanks.

Vshnv2001 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,109 @@
+     * @param tasks list of tasks to get number of existing tasks.
+     * @return response of duke in string.
+     */
+    public String showDeletedTask(Task taskToRemove, TaskList tasks) {
```
suggesting a change of method name to deletedTaskString- as the UI will be outputting the deleted task string to the console, this naming will help prevent confusion in the future.

Jonaspng replied:
Hmm. Makes sense thanks.

Jonaspng committed:
MainWindow Class: extracting out magic literals

Magic literals such as color of Dialog box and style of Dialog box were
identified. Magic literals makes code harder to understand.

Let's,
* change hex values for color into a constant
* change dialog box style into a constant

This make make our code more readable and easier to understand.

Jonaspng committed:
Remove unnecessary assert

Jonaspng committed:
Merge pull request #1 from Jonaspng/branch-A-Assertions

branch-A-Assertions

Jonaspng committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/MainWindow.java

Jonaspng committed:
Merge pull request #2 from Jonaspng/branch-A-CodeQuality

branch-A-CodeQuality

Jonaspng committed:
Add update function to change dates of task

Task's date could not be changed after it is registered into the system.

Allowing users to modify the date of existing tasks is more
user-friendly as users do not need to delete the task and add the
updated task again.

Let's,
* create a new command called updateDate
* modify parser to accept update command

Jonaspng committed:
Update java docs

Jonaspng committed:
Solve InvocationTargetException error

Jonaspng committed:
Upgrade search function to become non-case sensitive

Jonaspng committed:
Add personality to duke

Jonaspng committed:
Follow code reuse policy

Jonaspng committed:
Add Readme and Ui photo

Jonaspng committed:
Update README.md

Jonaspng committed:
Update SocDuke error messages

Jonaspng committed:
Update SocDuke test cases

Jonaspng committed:
Update README.md

Jonaspng committed:
Update build.gradle file

Jonaspng committed:
Update javadocs

Jonaspng committed:
Update Readme

Jonaspng committed:
Update java docs

Jonaspng committed:
Update README.md

Jonaspng committed:
Update java docs
